prov/lnx/src/lnx_av.c

```diff
@@ -0,0 +1,702 @@
+		if (rc)
+			frc = rc;
+	}
+
+	return frc;
+}
+
+static const char *
+lnx_av_straddr(struct fid_av *av, const void *addr,
+	       char *buf, size_t *len)
+{
+	/* TODO: implement */
+	return NULL;
+}
+
+static int
+lnx_av_lookup(struct fid_av *av, fi_addr_t fi_addr, void *addr,
+	      size_t *addrlen)
+{
+	/* TODO: implement */
+	return -FI_EOPNOTSUPP;
+}
+
+static struct fi_ops_av lnx_av_ops = {
+	.size = sizeof(struct fi_ops_av),
+	.insert = lnx_av_insert,
+	.remove = lnx_av_remove,
+	.insertsvc = fi_no_av_insertsvc,
+	.insertsym = fi_no_av_insertsym,
+	.lookup = lnx_av_lookup,
+	.straddr = lnx_av_straddr,
+};
+
+static void lnx_get_core_av_attr(struct local_prov_ep *ep,
+				 struct fi_av_attr *attr)
+{
+	memset(attr, 0, sizeof(*attr));
+	attr->type = ep->lpe_fi_info->domain_attr->av_type;
+}
+
+static int lnx_open_avs(struct local_prov *prov, struct fi_av_attr *attr,
+			void *context)
+{
+	int rc = 0;
+	struct local_prov_ep *ep;
+	struct fi_av_attr core_attr;
+
+	dlist_foreach_container(&prov->lpv_prov_eps,
+		struct local_prov_ep, ep, entry) {
+		lnx_get_core_av_attr(ep, &core_attr);
+		if (ep->lpe_local)
+			core_attr.count = ep->lpe_fi_info->domain_attr->ep_cnt;
+		else
+			core_attr.count = attr->count;
+		rc = fi_av_open(ep->lpe_domain, &core_attr,
+					    &ep->lpe_av, context);
+		if (rc)
+			return rc;
+	}
+
+	return 0;
+}
+
+int lnx_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
+		struct fid_av **av, void *context)
+{
+	struct lnx_fabric *fabric;
+	struct lnx_domain *lnx_domain;
+	struct lnx_peer_table *peer_tbl;
+	struct local_prov *entry;
+	size_t table_sz = LNX_DEF_AV_SIZE;
+	int rc = 0;
+
+	if (!attr)
+		return -FI_EINVAL;
+
+	if (attr->name)
+		return -FI_ENOSYS;
+
+	if (attr->type != FI_AV_UNSPEC &&
+	    attr->type != FI_AV_TABLE)
+		return -FI_ENOSYS;
+
+	if (attr->type == FI_AV_UNSPEC)
+		attr->type = FI_AV_TABLE;
+
+	peer_tbl = calloc(sizeof(*peer_tbl), 1);
+	if (!peer_tbl)
+		return -FI_ENOMEM;
+
+	if (attr->count != 0)
+		table_sz = attr->count;
+
+	peer_tbl->lpt_entries =
+	  calloc(sizeof(struct lnx_peer *) * table_sz, 1);
+	if (!peer_tbl->lpt_entries) {
+		rc = -FI_ENOMEM;
+		goto failed;
+	}
+
+	lnx_domain = container_of(domain, struct lnx_domain,
+				  ld_domain.domain_fid.fid);
+	fabric = lnx_domain->ld_fabric;
+
+	rc = ofi_av_init_lightweight(&lnx_domain->ld_domain, attr,
+				     &peer_tbl->lpt_av, context);
+	if (rc) {
+		FI_WARN(&lnx_prov, FI_LOG_CORE,
+			"failed to initialize AV: %d\n", rc);
+		goto failed;
+	}
+
+	peer_tbl->lpt_max_count = table_sz;
+	peer_tbl->lpt_domain = lnx_domain;
+	peer_tbl->lpt_av.av_fid.fid.ops = &lnx_av_fi_ops;
+	peer_tbl->lpt_av.av_fid.ops = &lnx_av_ops;
+
+	assert(fabric->lnx_peer_tbl == NULL);
+
+	/* need this to handle memory registration vi fi_mr_regattr(). We need
+	 * to be able to access the peer table to determine which endpoint
+	 * we'll be using based on the source/destination address */
+	fabric->lnx_peer_tbl = peer_tbl;
```

Comment:
I don't understand why this is needed for fi_mr_reg. For memory registration, seems like you only need the following:

- The lnx domain to maintain a list of all core domains. This is required to call fi_mr_reg() for each core provider.
- If lnx requires FI_MR_ENDPOINT, the lnx user should be providing the libfabric endpoint via fi_mr_bind().

I am not sure why AV is needed.